src/main/host/affinity.c

```diff
@@ -0,0 +1,74 @@
+#include "affinity.h"
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE // Make sure this is defined for CPU_* macros
+#endif              // _GNU_SOURCE
+
+#include <assert.h>
+#include <glib.h>
+#include <sched.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <sys/sysinfo.h>
+
+#include "main/core/support/options.h"
+#include "support/logger/logger.h"
+
+// Compute once and cache this value. Apparently fetching is expensive and
+// requires parsing procfs.
+static int _ncpus = -1; // Starts uninitialized.
+
+static gint _affinity_enabled = 0;
+
+OPTION_EXPERIMENTAL_ENTRY(
+    "cpu-pin", 0, 0, G_OPTION_ARG_INT, &_affinity_enabled,
+    "Pin workers to fixed CPU cores, pin plugins to the worker's core", "[0|1]")
+
+__attribute__((constructor)) static void _init_ncpus() { _ncpus = get_nprocs(); }
+
+int affinity_getGoodWorkerAffinity(guint worker_thread_id) {
+    assert(_ncpus > 0);                    // Our host better have at least one CPU.
+    return (int)worker_thread_id % _ncpus; // Uniform distribution of workers over CPUs
+}
+
+int affinity_setProcessAffinity(pid_t pid, int new_cpu_num, int old_cpu_num) {
+    assert(_ncpus > 0 && pid >= 0);
+
+    // We can short-circuit if there's no work to do.
+    if (!_affinity_enabled || new_cpu_num == AFFINITY_UNINIT || new_cpu_num == old_cpu_num) {
```

Comment:
If `new_cpu_num == AFFINITY_UNINIT` then don't we want to "unpin" the process to allow the scheduler to run it anywhere?

Reply:
I'm using `AFFINITY_UNINIT` to denote the case when AFFINITY is unknown or has never been set (as opposed to the case when affinity is set to all CPUs). I think it'd be better to introduce a new variable to unpin/"unset" affinity if we want that behavior.